### PR TITLE
Add Combine support for App and User

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

**Review by pavel-ship-it — APPROVED** [review]
LGTM.

File: RealmSwift/App.swift [hunk 1]
```diff
@@ -159,6 +159,81 @@ extension App {
...
+
+        return Self(app, callbackQueue: queue)
+    }
+}
+
+@available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, macCatalyst 13.0, macCatalystApplicationExtension 13.0, *)
+extension App: ObservableObject {
+    /// A publisher that emits Void each time the object changes.
+    ///
+    /// Despite the name, this actually emits *after* the collection has changed.
```
Comment (leemaguire) [inline RealmSwift/App.swift:231]: Is this doc correct?

File: RealmSwift/Sync.swift [hunk 2]
```diff
@@ -632,6 +632,68 @@ public extension User {
...
+
+        subscriber.receive(subscription: UserSubscription(user: user, token: token))
+    }
+}
+
+@available(OSX 10.15, watchOS 6.0, iOS 13.0, iOSApplicationExtension 13.0, OSXApplicationExtension 10.15, tvOS 13.0, macCatalyst 13.0, macCatalystApplicationExtension 13.0, *)
+extension User: ObservableObject {
+    /// A publisher that emits Void each time the object changes.
+    ///
+    /// Despite the name, this actually emits *after* the collection has changed.
```
Comment (leemaguire) [inline RealmSwift/Sync.swift:692]: Is this doc correct?

**Review by leemaguire — COMMENTED** [review]
Nice work! I've left some small comments.

File: RealmSwift/App.swift (hunk 1 above)
Comment (jsflax) [reply RealmSwift/App.swift:231]: No, will fix

File: Realm/RLMApp.mm [hunk 3]
```diff
@@ -375,6 +390,19 @@ - (void)authorizationController:(__unused ASAuthorizationController *)controller
...
+    RLMAppSubscriptionToken *token = [[RLMAppSubscriptionToken alloc] init];
+
+    token->_token = std::make_unique<realm::Subscribable<app::App>::Token>(_app->subscribe([block = std::move(block), self] (auto&) {
+        block(self);
+    }));
+    return token;
+}
+
+- (void)unsubscribe:(RLMAppSubscriptionToken *)token {
+    return _app->unsubscribe(*token->_token);
```
[the remark this answers is not in the document]
Comment (jsflax) [reply Realm/RLMApp.mm:403]: I'm not sure I agree with the visitor pattern here. Semantically, an App has subscribers and governs the subscription. Reversing that relationship seems like an odd choice.

File: Realm/RLMApp.mm [hunk 4]
```diff
@@ -375,6 +390,19 @@ - (void)authorizationController:(__unused ASAuthorizationController *)controller
     [self.authorizationDelegate authenticationDidCompleteWithError:error];
 }
 
+- (RLMAppSubscriptionToken *)subscribe:(RLMAppNotificationBlock)block {
+    RLMAppSubscriptionToken *token = [[RLMAppSubscriptionToken alloc] init];
+
+    token->_token = std::make_unique<realm::Subscribable<app::App>::Token>(_app->subscribe([block = std::move(block), self] (auto&) {
```
[the remark this answers is not in the document]
Comment (jsflax) [reply Realm/RLMApp.mm:396]: That could work here, yes.

File: Realm/RLMApp.mm [hunk 5]
```diff
@@ -214,6 +217,18 @@ - (void)setDefaultRequestTimeoutMS:(NSUInteger)defaultRequestTimeoutMS {
...
 }
 
+#pragma mark RLMAppSubscriptionToken
+@implementation RLMAppSubscriptionToken {
+@public
+    std::unique_ptr<realm::Subscribable<app::App>::Token> _token;
+}
+
+- (NSUInteger)value {
+    return _token->value();
```
[the remark this answers is not in the document]
Comment (jsflax) [reply Realm/RLMApp.mm:227]: No.

File: Realm/Realm.modulemap [hunk 6]
```diff
@@ -6,6 +6,7 @@ framework module Realm {
 
     explicit module Private {
         header "RLMAccessor.h"
+        header "RLMApp_Private.h"
```
[the remark this answers is not in the document]
Comment (jsflax) [reply Realm/Realm.modulemap:9]: Good catch, done.

**Review by leemaguire — APPROVED** [review]

**Comment by tgoyne** [comment]
The swiftlint failure isn't a false-positive.

**Review by tgoyne — APPROVED** [review]

**Pull request merged**